# Binomial Distribution using the Python Math Library


The binomial distribution can be used to model real world scenarios where the events can be classified as one of two possible outcomes (success or failure). Using the binomial we can calculate the probability of obtaining a given number of success (x) from a fixed number of trials (n), with a probability (p) of success on each trial that does not change and the outcome of each trial is independent of all previous trials.

In this notebook I will use the binomial distribution to calculate probabilities. Specifically, I will use the Probability Mass Function for the binomial and the Cumulative Distribution Function to obtain probabilities. On this occasion I will make use of the python Math library to calculate and wirk with the binomial distribution.

In [1]:
# Importing the math software library. 

import math

The math libary does not have a function for the binomial distribution directly but it contains useful function that are needed to calculate the binomial. It contains a function called comb which gives us the binomial coefficient (n choose k: From n trials choose k number of successes). This makes life easier as it allows us to use the comb function when defining a function that calculates the Probability Mass Function (pmf) for the binomial. 

In [2]:
# Defining a function for the binomial pmf, making use of the comb function. Comb is likely an abbreviation for combinatorial.
def binom_pmf(n, x, p):
    return math.comb(n, x) * p**x * (1-p)**(n-x)

In [3]:
# Testing the binomial_pmf function with 4 successes from 10 trials and a success probability per trial of 0.3. 

binom_pmf(10, 4, 0.3)

0.2001209489999999

We can see above that the probability of getting four successes from 10 trials, when p = 0.3, is about 0.2 or 20%.

To obtain the binomial Cumulative Distribution Function (cdf) we need to sum over all values of the pmf. To do this we need to make use of a floor operator. This is a mathematical operation that rounds positive decimals down to the nearest integer value and negative decimals away from zero. The Math library has a floor function that we can make use of to do this. 

In [4]:
# Defining a function for the binomial cdf. Making use of the floor method contained within the math library. 

def binom_cdf(n, x, p):
    p_val = 0
    x_floor = math.floor(x)
    for i in range(x_floor+1):
        p_val = p_val + binom_pmf(n, i, p)
    return p_val

In [5]:
binom_cdf(10, 4, 0.3)

0.8497316673999995

The cdf for the binomial gives us the probability of getting x or fewer successes from n trials. In this case, the probability of getting 4 or fewer successes from 10 trials, with a success probability on each trial of 0.3, is about 0.85 or 85%. 